nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -8,21 +8,183 @@ use super::DataStore;
+                current_version
+            }
+            Err(e) => {
+                return Err(format!("Cannot read schema version: {e}"));
+            }
+        };
+
+        let Some(config) = config else {
+            return Err("Not configured to automatically update schema".to_string());
+        };
+
+        if current_version > desired_version {
+            return Err("Nexus older than DB version: automatic downgrades are unsupported".to_string());
+        }
+
+        // If we're here, we know the following:
+        //
+        // - The schema does not match our expected version (or at least, it
+        // didn't when we read it moments ago).
+        // - We should attempt to automatically upgrade the schema.
+        //
+        // We do the following:
+        // - Look in the schema directory for all the changes, in-order, to
+        // migrate from our current version to the desired version.
+
+        info!(log, "Reading schemas from {}", config.schema_dir.display());
+        let mut dir = tokio::fs::read_dir(&config.schema_dir)
+            .await
+            .map_err(|e| format!("Failed to read schema config dir: {e}"))?;
+        let mut all_versions = BTreeSet::new();
+        while let Some(entry) = dir
+            .next_entry()
+            .await
+            .map_err(|e| format!("Failed to read schema dir: {e}"))?
+        {
+            if entry.file_type().await.map_err(|e| e.to_string())?.is_dir() {
+                let name = entry
+                    .file_name()
+                    .into_string()
+                    .map_err(|_| "Non-unicode schema dir".to_string())?;
+                if let Ok(observed_version) = name.parse::<SemverVersion>() {
+                    all_versions.insert(observed_version);
+                } else {
+                    warn!(log, "Failed to parse {name} as a semver version");
+                }
+            }
+        }
+
+        if all_versions.get(&current_version).is_none() {
+            return Err(format!(
+                "Current DB version {current_version} is not known"
+            ));
+        }
+
+        let target_versions = all_versions.range((
+            Bound::Excluded(&current_version),
+            Bound::Included(&desired_version),
+        ));
+
+        for target_version in target_versions.into_iter() {
+            info!(
+                log,
+                "Attempting to grab lease for upgrade to version {}",
+                target_version
+            );
+
+            // Before we attempt to perform an update, grab a lease to allow
+            // this Nexus instance to perform the update.
+            //
+            // If another instance is performing an update, we will fail, and
+            // let them to the work on our behalf.
+            let lease = self
+                .grab_schema_update_lease(&current_version, &target_version)
+                .await
+                .map_err(|e| e.to_string())?;
```

Comment:
Anytime I hear the word "lease" I get concerned, because of the rough edges involved in making it work. I think you are making the right decision short term on not expiring the lease based on time as that alleviates the bulk of things that can go wrong from a safety POV. You can only lose liveness if the leas holder crashes, as you say. 

If we end up sticking with this, and expiring leases based on time then we must use a fencing token to gate an old lease holder who doesn't realize it has expired from writing over a newer lease holder. It's unclear to me if the `current_version` filter is good enough here. It may be that it is fine because if it matches the stale lease holder is just going to try the same migration and fail. 

This is my go to resource for discussing distributed leases/locks and fencing tokens: https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html

You can ignore all the redlock stuff. Important sections are "Protecting a resource with a lock" and "Making the lock safe with fencing"

Reply:
I spent a while working on this, and I have an alternate approach.

https://www.cockroachlabs.com/docs/stable/online-schema-changes#schema-changes-within-transactions

- CockroachDB cautions strongly against performing multiple DDL statements in a transaction
- CockroachDB recommends strongly against performing both DDL and DML statements in a transaction (since one might succeed, and the other might fail).

However, they **don't** disallow running a **query** operation in a transaction which contains a single DDL operation.

So, I have the following proposal for upgrading from N -> N+1:

1. Mark the schema upgrade as in-progress, if we are at the version we expect to be:

```sql
UPDATE db_metadata SET target_version = N + 1 WHERE version = N AND singleton = true;
```

2. Perform the schema upgrade in a transaction, confirming that the version and target versions have not changed:

```sql
BEGIN;

SELECT CAST(
    IF(
        (
            SELECT version = N and target_version = N + 1
            FROM omicron.public.db_metadata WHERE singleton = true
        ),
        'true',
        'Invalid starting version for schema change'
    ) AS BOOL
);

ALTER TABLE omicron.public.instance
    ADD COLUMN IF NOT EXISTS boot_on_fault BOOL NOT NULL DEFAULT false;

COMMIT;
```

3. Finalize the update

```sql
UPDATE db_metadata SET version = N+1, target_version = NULL WHERE version = N AND target_version = N +1 AND singleton = true;
```

These steps are safe to execute from multiple Nexus instances concurrently, and they abandon the DDL changes which might be taken if a different Nexus manages to complete the upgrade early.

I also was able to write a test for this behavior, where I have the following versions:

- v0: Starting version
- v1: Sleep a random amount, Add table
- v2: Delete table

And I do the following:

- Launch 10 Nexus Instances
- Let them all try to upgrade from v0 -> v1 -> v2
- Observe that, as each of them finishes initializing the datastore, they **cannot see the table, because it has been deleted, and no "late upgrades" re-add it**.

(Without the transactional query in step (2), this test fails. With it, this test succeeds)

Reply:
Why I like this new version:
- There is no "lease" or "lock" mechanism necessary. Nexus instances can concurrently execute.
- If Nexus instances crash before step (3), they should be able to re-do steps (1) and (2) to ensure the update was applied.
- As soon as step (3) is applied, the update is complete, and subsequent calls to step (1) or (2) will fail (as they should!)

Reply:
Got this feedback from a member of Cockroach Labs, when I asked about the safety the transaction in step (2):

> Hello! The code for a transaction that you showed is safe — it will block the DDL if there’s a concurrent UPDATE happening. Just to note that the hazard is from performing multiple schema changes within a single explicit transaction, so you may want to avoid running other DDL/DML after the ALTER TABLE  transactionally.